### PR TITLE
Fix fatal error in purchasing a subscription with PRB and ECE when the subscription's price or sign-up fee is a string

### LINKED ISSUE #2966: Fatal error with PRBs and Subscriptions when the subscription's price or sign-up fee is a string
**Describe the bug**
There's a fatal error when using Payment Request Buttons, like Google Pay or Apple Pay, with subscriptions in some scenarios.

The error comes from [this line](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/8.0.0/includes/payment-methods/class-wc-stripe-payment-request.php#L365).

WC_Subscriptions_Product::get_sign_up_fee() [may return a string](https://github.com/Automattic/woocommerce-subscriptions-core/blob/trunk/includes/class-wc-subscriptions-product.php#L526-L535).

<details>

```
PHP Fatal error: Uncaught TypeError: Unsupported operand types: string + string in .../wp-content/plugins/woocommerce-gateway-stripe/includes/payment-methods/class-wc-stripe-payment-request.php:365
Stack trace:
#0 .../wp-content/plugins/woocommerce-gateway-stripe/includes/payment-methods/class-wc-stripe-payment-request.php(416): WC_Stripe_Payment_Request->get_product_price(Object(WC_Product_Subscription_Variation))
#1 .../wp-content/plugins/woocommerce-gateway-stripe/includes/payment-methods/class-wc-stripe-payment-request.php(806): WC_Stripe_Payment_Request->get_product_data()
#2 .../wp-content/plugins/woocommerce-gateway-stripe/includes/payment-methods/class-wc-stripe-payment-request.php(839): WC_Stripe_Payment_Request->javascript_params()
#3 .../wp-includes/class-wp-hook.php(324): WC_Stripe_Payment_Request->scripts('')
#4 .../wp-includes/class-wp-hook.php(348): WP_Hook->apply_filters(NULL, Array)
#5 .../wp-includes/plugin.php(517): WP_Hook->do_action(Array)
#6 .../wp-includes/script-loader.php(2262): do_action('wp_enqueue_scri...')
#7 .../wp-includes/class-wp-hook.php(324): wp_enqueue_scripts('')
#8 .../wp-includes/class-wp-hook.php(348): WP_Hook->apply_filters(NULL, Array)
#9 .../wp-includes/plugin.php(517): WP_Hook->do_action(Array)
#10 .../wp-includes/general-template.php(3052):
```
</details>

**To Reproduce**
TBC (I think this happens with PHP 8.1)

**Expected behavior**
No errors must be triggered when using PRBs with subscriptions.

**Additional context**

Slack thread p1709233934325419-slack-C7U3Y3VMY

### PR DESCRIPTION
**Pull request by Mayisha**
Fixes #2966 

In `get_product_price` function for PRB/ECE we need `$product->get_price()`, `wc_get_price_including_tax`, `wc_get_price_excluding_tax`, `WC_Subscriptions_Product::get_sign_up_fee` for different scenarios. These functions may return the result in different types as per the corresponding function definitions `int`/`float`/`string`. 

When the types are different the following code causes a fatal error-

```
$product_price = $product->get_price() + WC_Subscriptions_Product::get_sign_up_fee( $product );
```

## Changes proposed in this Pull Request:
- casting the prices to `float` to avoid the fatal error.

## Testing instructions
- The error is not always reproducible. So we should test subscription purchase 
- Create a few products if you don't have them already-
    - A subscription product with no signup fee
    - A variable subscription product with a signup fee
    - A simple subscription product with a signup fee
- Purchase these type of products with PRB button. Confirm that there is no fatal error.
- Enable the ECE feature flag.
- Purchase these type of products with ECE button. Confirm that there is no fatal error.

### CODE REVIEW

File: includes/payment-methods/class-wc-stripe-express-checkout-helper.php [hunk 1]
```diff
@@ -147,7 +147,7 @@ public function get_total_label() {
 	 * @param bool|null $is_deposit      Whether this is a deposit.
 	 * @param int       $deposit_plan_id Deposit plan ID.
 	 *
-	 * @return integer Total price.
+	 * @return int|float Total price.
 	 */
 	public function get_product_price( $product, $is_deposit = null, $deposit_plan_id = 0 ) {
```
Comment (wjrosa) [inline includes/payment-methods/class-wc-stripe-express-checkout-helper.php:152]: Hum, what if we keep it more consistent/strict, always returning the same type? Either we cast anything to float or int when returning

Comment (Mayisha) [reply includes/payment-methods/class-wc-stripe-express-checkout-helper.php:152]: Addressed in [3162320](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3617/commits/3162320d37fc5709cfe6eb38f0c567d4fa9457f6)

File: includes/payment-methods/class-wc-stripe-payment-request.php [hunk 2]
```diff
@@ -366,16 +366,16 @@ public function get_button_label() {
 	 * @since 5.2.0
 	 *
 	 * @param object $product WC_Product_* object.
-	 * @return integer Total price.
+	 * @return float Total price.
 	 */
 	public function get_product_price( $product ) {
 		$product_price = $product->get_price();
```
Comment (wjrosa) [inline includes/payment-methods/class-wc-stripe-payment-request.php]:
```suggestion
		$product_price = (float) $product->get_price();
```

File: includes/payment-methods/class-wc-stripe-payment-request.php [hunk 3]
```diff
@@ -366,16 +366,16 @@ public function get_button_label() {
...
 		// Add subscription sign-up fees to product price.
 		if ( in_array( $product->get_type(), [ 'subscription', 'subscription_variation' ] ) && class_exists( 'WC_Subscriptions_Product' ) ) {
-			$product_price = $product->get_price() + WC_Subscriptions_Product::get_sign_up_fee( $product );
+			$product_price = (float) $product_price + (float) WC_Subscriptions_Product::get_sign_up_fee( $product );
 		}
 
-		return $product_price;
+		return (float) $product_price;
```
Comment (wjrosa) [inline includes/payment-methods/class-wc-stripe-payment-request.php]:
```suggestion
		return $product_price;
```

File: includes/payment-methods/class-wc-stripe-payment-request.php [hunk 4]
```diff
@@ -366,16 +366,16 @@ public function get_button_label() {
...
 	 * @param object $product WC_Product_* object.
-	 * @return integer Total price.
+	 * @return float Total price.
 	 */
 	public function get_product_price( $product ) {
 		$product_price = $product->get_price();
 		// Add subscription sign-up fees to product price.
 		if ( in_array( $product->get_type(), [ 'subscription', 'subscription_variation' ] ) && class_exists( 'WC_Subscriptions_Product' ) ) {
-			$product_price = $product->get_price() + WC_Subscriptions_Product::get_sign_up_fee( $product );
+			$product_price = (float) $product_price + (float) WC_Subscriptions_Product::get_sign_up_fee( $product );
```
Comment (wjrosa) [inline includes/payment-methods/class-wc-stripe-payment-request.php]:
```suggestion
			$product_price += (float) WC_Subscriptions_Product::get_sign_up_fee( $product );
```

**Review by wjrosa — APPROVED** [review]
Nice one, Mayisha. Works as intended!

I left some optional suggestions (which I believe makes the code easier to read).

**Comment by mdrovdahl** [comment]
👏 I just ran across this today on a subscription product variation that had an empty Price field. It would cause a fatal type error in `get_product_price()`.

**Pull request merged**